crates/circuit/src/operations.rs

```diff
@@ -2046,7 +2046,31 @@ fn multiply_param(param: &Param, mult: f64, py: Python) -> Param {
     }
 }
 
-fn add_param(param: &Param, summand: f64, py: Python) -> Param {
+pub fn rmultiply_param(param1: Param, param2: Param, py: Python) -> Param {
+    match [param1, param2] {
+        [Param::Float(theta), Param::Float(lambda)] => Param::Float(theta * lambda),
+        [Param::ParameterExpression(theta), Param::Float(lambda)] => Param::ParameterExpression(
+            theta
+                .clone_ref(py)
+                .call_method1(py, intern!(py, "__rmul__"), (lambda,))
+                .expect("Parameter expression multiplication failed"),
+        ),
+        [Param::Float(theta), Param::ParameterExpression(lambda)] => {
+            rmultiply_param(Param::ParameterExpression(lambda), Param::Float(theta), py)
+        }
+        [Param::ParameterExpression(theta), Param::ParameterExpression(lambda)] => {
+            Param::ParameterExpression(
+                theta
+                    .clone_ref(py)
+                    .call_method1(py, intern!(py, "__rmul__"), (lambda,))
+                    .expect("Parameter expression multiplication failed"),
+            )
+        }
+        _ => unreachable!("Unsupported multiplication."),
+    }
+}
+
+pub fn add_param(param: &Param, summand: f64, py: Python) -> Param {
```

Comment:
A small suggestion at this point: consider something like this instead to remove code duplication (the `call_method1` block and the recursion):
```rust
pub fn rmultiply_param(param1: Param, param2: Param, py: Python) -> Param {
    let (p1, p2) = match (&param1, &param2) {
        (Param::Float(theta), Param::Float(lambda)) => return Param::Float(theta * lambda),
        (Param::ParameterExpression(_), Param::Float(_)) => (param1, param2),
        (Param::Float(_), Param::ParameterExpression(_)) => (param2, param1),
        (Param::ParameterExpression(_), Param::ParameterExpression(_)) => (param1, param2),
        _ => unreachable!("Unsupported multiplication."),
    };

    let Param::ParameterExpression(op1) = p1 else {unreachable!("Param::ParameterExpression is assumed")};

    Param::ParameterExpression(
    op1
    .clone_ref(py)
    .call_method1(py, intern!(py, "__rmul__"), (p2,))
    .expect("Parameter expression multiplication failed")
    )
}
```

Reply:
I refactored the function in 71ca90b7b 🙂 I applied a slightly different version that above, which simply calls `multiply_param(Param, f64)` for the two mixed cases of `ParameterExpression, Float`. So this even further simplifies the code and also gets rid of the duplication 👍🏻